app/commands/commands.go

```diff
@@ -0,0 +1,11 @@
+package commands
+
+// Command defines the interface for each command that is to be defined
+type Command interface {
+	// Valid method returns true/false based on whether it has valid data for execution
+	Valid() bool
```

Comment:
I think it shouldn't be possible to construct an invalid command. This can be achieved by performing validation before creating the command.

This would mean the `Valid()` method would no longer exist, and there would be no need for the `unrecognized` command.

Reply:
I think the [Command Pattern](https://en.wikipedia.org/wiki/Command_pattern) would be appropriate here, and this PR (along with the changes in https://github.com/Automattic/neurobot/commit/31a495445bf152710ec5a4dd2ede89f409473d74) are close to an implementation of that pattern.

The command pattern can be implemented with a *command bus* which is basically the same thing as event bus: you publish something to the bus, listeners receive it. There are some differences between the two though:

1. An event is something that happened, a command is an action to be taken. So:
2. An event can be handled by multiple listeners, a command most probably should be handled only once.

Internally a command bus can use an event bus, but it should guarantee that only one listener will handle the event. In the code in https://github.com/Automattic/neurobot/commit/31a495445bf152710ec5a4dd2ede89f409473d74, the event bus is the channel, and IMHO that's fine but you could use [the event bus](https://github.com/Automattic/neurobot/blob/master/infrastructure/event/bus.go) instead.

----

I would implement the command pattern here as follows:

## CommandHandler
Provides an interface all command handlers must implement. Each command type will have it's own handler, e.g. the `DrinkWater` command will have a `DrinkWaterHandler`, the `EatFood` command will have an `EatFoodHandler`.

```go
type CommandHandler interface {
    Handle(command interface{})
}
```

## CommandBus
This is where commands are published to, and where handlers can be registered. So two methods are needed:

```go
type CommandBus interface {
    // Or Send, or whatever other similar term
    Publish(c *Command)

    // Register a handler for a command type.
    // Throws an error if a handler is already registered for that command type.
    Register(commandType string, handler CommandHandler) Error
```

The instance of command bus can then be injected into whoever needs to publish commands, or register command handlers.

## Validation
It should not be possible to create an invalid command. This means that the `Publish()` method on the command bus should error out if the command is invalid, which will give the publisher of the command a chance to gracefully handle the error.

```go
type CommandBus interface {
    Publish(c *Command) Error
}
```

## `registry`
Currently it's the bot `registry` that listens to messages from matrix, so it will need the instance of the command bus. If `Publish()` returns an error, the registry will write something back to matrix, e.g. usage hints. 

## `app`
`app` is the entry point for workflows coming from webhooks, so it would make sense to make it the single entry point, and also have it "listen" for commands. So `app` will need the `CommandBus` instance so it can call `Register()`.

Reply:
@psrpinto Thank you for the detailed response! I will study this on my personal time.